# Blog Post Summarizer
### Using Hugging Face Transformers to summarize blog posts

This notebook demonstrates how to use a pre-trained NLP model (Facebook's BART) 
to automatically generate summaries of blog posts.

**Model:** `facebook/bart-large-cnn`  
**Task:** Abstractive text summarization  
**Author:** Your Name  

---

In [2]:
# Core libraries
from transformers import pipeline
import textwrap
import os

print("Libraries imported successfully")

The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.


0it [00:00, ?it/s]

Libraries imported successfully


## Step 1: Load the Summarization Model

We're using Facebook's BART model, pre-trained on a large corpus of news articles 
and fine-tuned specifically for summarization. 

The first time this runs it will download the model (~1.6GB) and cache it. 
Every run after that will load instantly from the cache.

In [3]:
print("Loading model... (this may take a few minutes the first time)")

summarizer = pipeline("summarization", model="facebook/bart-large-cnn")

print("Model loaded successfully!")

Loading model... (this may take a few minutes the first time)


/opt/conda/lib/python3.11/site-packages/huggingface_hub/file_download.py:949: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Model loaded successfully!


## Step 2: Load a Blog Post

We'll start by defining a sample blog post as a string directly in the notebook.
Once we confirm the model is working correctly, we'll move on to loading posts 
from actual text files.

In [4]:
sample_post = """
There are gaggles of geese.

A group of bears is referred to as a sloth of bears.

There are the well-known teams of oxen, hives of bees, and packs of wolves.

We even collectively group people together.

There are companies of actors. Bands of musicians. Regiments of soldiers.

Yet there are no generally well-known listings for Engineers.

Go ahead, look it up. After spending more than a few minutes of research here there is shockingly little in the way of generally agreed upon group nomenclature for some of the finest groups of humans to ever walk this earth.

It is possible to fall back on the possibly indelicate and often scatological urban dictionary which refers to a group of more than three engineers together as:

A wank of engineers.

Look that one up at your own risk.

Yet, jokes regarding this collectively talented group of people abound. It is a well-regarded profession. Why shouldn’t we have some collective nouns to fall back on when we gather?

It isn’t like we take up a lot of space! Haven’t you heard the story about the Engineers taking a train ride?

Once upon a time three physicists and three Engineers found themselves on a train platform heading to a big conference. The physicists watched in confusion as only one Engineer went to the ticket window and purchased a ticket.

After a few minutes, they all started boarding the train. A few minutes later it left and the conductor started collecting tickets. The physicists were amazed to see all three Engineers get up and crowd into the bathroom on the car. As the conductor passed by, he knocked on the bathroom door asking for tickets.

A single arm with a single ticket emerged, whereupon the conductor accepted their ticket and carried on.

If this isn’t an intelligent and space-conserving method for travel, then I don’t know what is.

So what descriptive, yet informative appellation should we place upon these enterprising geniuses?

The following is an attempt to categorize and collectively identify these amazing groups of smart and talented people. As there appear to be no globally recognized official terms for groups of these people, here is my offering to the Engineering Gods describing collective nouns for different branches of Engineering.

What Do You Call A Group Of Chemical Engineers?

We start the collective naming by trying to identify this coagulation of chemical junkies. Dealing with molecules, materials, and moles, chemical engineers convert raw materials into usable stuff that we use and rely on every single day.

Therefore, to honor the grouping of basic elemental units that together form complex and intricate supergroups, chemical engineers are collectively:
A Compound of Engineers

Take one of these groups and stick them in a battery and you have enough juice to move the world.

What Do You Call A Group Of Civil Engineers?

Lend me your roads, rails, and bridges! To the people that start at the bedrock of our society and build the structures and systems that keep us moving, safe, and healthy there are a lot of great possibilities here.

There are a lot of sub-disciplines to this group. We could call them boring for the tunnel engineers. Blueprints cover a bunch of other sub-disciplines, from structural engineers to municipal engineering. It is probably safe skipping over the sewage-based group names for environmental engineers.

However, one name rises to the top here. Or, maybe a more appropriate phrasing would be “settles to the bottom” and represents the core contributions by this group of people. Therefore, a group of Civil Engineers is:
A Foundation of Engineers

It would be remiss of us to not take a moment, or even a second moment to reflect on the shear joy of this conservative force of talented engineers who keep our world running smooth day in and day out.

Although, if you stuff too many Civil Engineers into a room they might end up being decidedly uncivil…

What Do You Call A Group Of Electrical Engineers?

There is a lot of potential for this group! The sparkies are lining up for this one. The term node might work here but then there would forever be arguments over the select few that want to be called supernodes, so that one is out.

Even the Simpsons contributed a worthy contender as the local baseball team is called the isotopes.

However, before we short circuit on names here, let’s charge things up with the name of a group of Electrical Engineers:
A Plug of Engineers

This one is more for the rest of the world than it is for the actual group, and it reflects some of the most basic tenants in computing, electronics, and electrical systems:

Is it plugged in?

And

Did you unplug it and then plug it back in?

A good group name provides easy context for the uninitiated and for those in the group itself. Besides, at the end of the day, aren’t we all just plugging things in?

What Do You Call A Group Of Mechanical Engineers?

This one is tough. A bearing of engineers sounds good and so does a tolerance of engineers. There are cars, trains, and planes here as well as motors, thermodynamics, and robotics.

However, as this group deals with mechanical motion in much of the work done and the ubiquitous symbol for this
A Gearing of Engineers

Let’s hope that any of these groups mesh together to form a constructive and cooperating machine. Besides, unlike the motors that they design and build, this collective group usually has an output higher than the sum of the individual parts.

What Do You Call A Group Of Software Engineers?

This is a tough one. It seems like it should be easy as Software Engineering contains a great many concepts related to gathering together a group of something. Arrays, collections, blocks, and many other concepts define concept groupings in this branch of Engineering. There is one name that stands out from the rest:
A Stack of Engineers

However, the concept of a stack wins out here. This term has grown to be rather large and wide-ranging in software development. It encompasses different ideas based on the context. It can define a data structure. It also can describe the collective technologies that make up the different components of a solution. One of the most famous sources for software development knowledge is known as Stack Exchange.

So this is a great collective noun to represent this group of industrious and enterprising wrangles of the 1s and 0s of our world. All hail the stacks.
What Do You Call A Group Of Applied Engineers?

This group of people deals with Applied Engineering. According to Wikipedia, they deal with “Automation/control systems/mechatronics/robotics“. Therefore, they get one of the best group names on here. A group of Applied Engineers is:
A Goldberg of Engineers

For the uninitiated, this is a reference to the wonderful Rube Goldberg who is known for creating cartoons with overly complex and intricate machines designed to only complete simple tasks.

As for the three engineers on the train? Their story didn’t end after their single ticket was punched.

After an informative and successful conference, the Engineers and physicists again found themselves on a train platform ready to board a train for the ride home.

The physicists, feeling like they were at least equal to the Engineers, proceeded to send a member up to the window to purchase only a single ticket.

However, to the consternation and confusion of the physicists, the Engineers did not purchase any tickets and remained quietly to the side, watching the actions of the three physicists.

This confusion only grew after boarding, as when the train started both the Engineers and physicists crowded into bathrooms on the train car.

It was only after everyone was settled in their cramped quarters that an Engineer emerged from their bathroom to walk over to the other bathroom, knock quietly on the door and ask:

“Tickets please!“

Thanks for reading!
"""

# Show basic stats about the post
word_count = len(sample_post.split())
print(f"Post loaded successfully")
print(f"Word count: {word_count}")
print(f"Character count: {len(sample_post)}")
print(f"\nFirst 200 characters preview:")
print(sample_post[:100])

Post loaded successfully
Word count: 1331
Character count: 7889

First 200 characters preview:

There are gaggles of geese.

A group of bears is referred to as a sloth of bears.

There are the well-known teams of oxen, hives of bees, and packs of wolves.

We even collectively group people toget


## Step 3: Summarize the Blog Post

BART has a maximum input of 1024 tokens (roughly 750-800 words). Our post is 
1331 words so we need to truncate it before passing it to the model.

This is a real-world constraint worth understanding — in a later step we'll 
explore chunking as a more sophisticated solution.

In [5]:
# BART has a max input of 1024 tokens (~750-800 words)
# We'll truncate for now and revisit chunking later
max_input_chars = 3000

truncated_post = sample_post[:max_input_chars]

print(f"Original character count: {len(sample_post)}")
print(f"Truncated character count: {len(truncated_post)}")
print("\nGenerating summary...")

summary = summarizer(
    truncated_post,
    max_length=150,
    min_length=50,
    do_sample=False
)

print("\n--- SUMMARY ---\n")
print(textwrap.fill(summary[0]['summary_text'], width=80))

Original character count: 7889
Truncated character count: 3000

Generating summary...

--- SUMMARY ---

There are no generally well-known listings for Engineers. A group of bears is
referred to as a sloth of bears. There are companies of actors. Bands of
musicians. Regiments of soldiers. Chemical engineers are collectively known as
chemical junkies.


In [6]:
print(sample_post[:3000])


There are gaggles of geese.

A group of bears is referred to as a sloth of bears.

There are the well-known teams of oxen, hives of bees, and packs of wolves.

We even collectively group people together.

There are companies of actors. Bands of musicians. Regiments of soldiers.

Yet there are no generally well-known listings for Engineers.

Go ahead, look it up. After spending more than a few minutes of research here there is shockingly little in the way of generally agreed upon group nomenclature for some of the finest groups of humans to ever walk this earth.

It is possible to fall back on the possibly indelicate and often scatological urban dictionary which refers to a group of more than three engineers together as:

A wank of engineers.

Look that one up at your own risk.

Yet, jokes regarding this collectively talented group of people abound. It is a well-regarded profession. Why shouldn’t we have some collective nouns to fall back on when we gather?

It isn’t like we take up a 

In [7]:
# Test truncating from middle instead of beginning
middle_start = len(sample_post) // 4
test_chunk = sample_post[middle_start:middle_start + 3000]

print("Generating summary from middle section...")

test_summary = summarizer(
    test_chunk,
    max_length=150,
    min_length=50,
    do_sample=False
)

print("\n--- SUMMARY (middle section) ---\n")
print(textwrap.fill(test_summary[0]['summary_text'], width=80))

Generating summary from middle section...

--- SUMMARY (middle section) ---

No globally recognized official terms for groups of these people. Here is my
offering to the Engineering Gods describing collective nouns for different
branches of Engineering. What Do You Call A Group Of Chemical Engineers? We
start the collective naming by trying to identify this coagulation of chemical
junkies. A group of Civil Engineers is:                A Foundation of
Engineers.


## Step 4: Chunking — Handling Long Documents Properly

Naively truncating a post from the beginning produces poor summaries because 
blog posts don't follow the "inverted pyramid" structure that BART was trained on.

The solution is **chunking**:
1. Split the full post into overlapping sections
2. Summarize each section independently  
3. Combine those summaries into a final summary

This is called map-reduce summarization and is the standard approach for 
documents that exceed a model's token limit.

In [8]:
def chunk_text(text, chunk_size=1000, overlap=100):
    """
    Split text into overlapping chunks.
    
    Args:
        text: The full text to chunk
        chunk_size: Characters per chunk
        overlap: Characters of overlap between chunks
                 (prevents losing context at boundaries)
    
    Returns:
        List of text chunks
    """
    chunks = []
    start = 0
    
    while start < len(text):
        end = start + chunk_size
        chunk = text[start:end]
        chunks.append(chunk)
        start = end - overlap  # step back by overlap amount
    
    return chunks

# Split our post into chunks
chunks = chunk_text(sample_post, chunk_size=1000, overlap=100)

print(f"Post split into {len(chunks)} chunks")
for i, chunk in enumerate(chunks):
    print(f"Chunk {i+1}: {len(chunk)} characters")

Post split into 9 chunks
Chunk 1: 1000 characters
Chunk 2: 1000 characters
Chunk 3: 1000 characters
Chunk 4: 1000 characters
Chunk 5: 1000 characters
Chunk 6: 1000 characters
Chunk 7: 1000 characters
Chunk 8: 1000 characters
Chunk 9: 689 characters


## Step 5: Summarize Each Chunk

Now we run each chunk through BART independently, then combine those 
individual summaries into a single final summary. This ensures every 
part of the post is represented.

In [9]:
def summarize_chunks(chunks, summarizer):
    """
    Summarize each chunk independently.
    
    Args:
        chunks: List of text chunks
        summarizer: The Hugging Face pipeline
    
    Returns:
        List of summaries, one per chunk
    """
    chunk_summaries = []
    
    for i, chunk in enumerate(chunks):
        print(f"Summarizing chunk {i+1} of {len(chunks)}...")
        
        # Skip chunks that are too short to summarize meaningfully
        if len(chunk.split()) < 30:
            print(f"  Chunk {i+1} too short, skipping")
            continue
            
        summary = summarizer(
            chunk,
            max_length=60,
            min_length=20,
            do_sample=False
        )
        
        chunk_summaries.append(summary[0]['summary_text'])
        print(f"  Done: {summary[0]['summary_text'][:80]}...")
    
    return chunk_summaries

chunk_summaries = summarize_chunks(chunks, summarizer)

print(f"\n✓ Generated {len(chunk_summaries)} chunk summaries")

Summarizing chunk 1 of 9...
  Done: There are no generally well-known listings for Engineers. It is possible to fall...
Summarizing chunk 2 of 9...
  Done: Three physicists and three Engineers were on a train to a big conference. They w...
Summarizing chunk 3 of 9...
  Done: The following is an attempt to categorize and collectively identify these amazin...
Summarizing chunk 4 of 9...
  Done: What Do You Call A Group Of Civil Engineers? A Foundation of Engineers. A Group ...
Summarizing chunk 5 of 9...
  Done: What Do You Call A Group Of Electrical Engineers? The sparkies are lining up for...
Summarizing chunk 6 of 9...
  Done: A good group name provides easy context for the uninitiated and for those in the...
Summarizing chunk 7 of 9...
  Done: Software Engineering contains a great many concepts related to gathering togethe...
Summarizing chunk 8 of 9...
  Done: A group of Applied Engineers is: a. noun to represent this group of industrious ...
Summarizing chunk 9 of 9...
  Done: The 

## Step 6: Combine Chunk Summaries into a Final Summary

With each chunk summarized individually, we now concatenate those summaries 
and run them through BART one final time to produce a single coherent summary 
of the entire post.

This two-pass approach (summarize chunks → summarize the summaries) is the 
core of map-reduce summarization.

In [10]:
def final_summary(chunk_summaries, summarizer):
    """
    Combine chunk summaries into one final summary.
    
    Args:
        chunk_summaries: List of individual chunk summaries
        summarizer: The Hugging Face pipeline
    
    Returns:
        Final combined summary string
    """
    # Join all chunk summaries into one block of text
    combined = " ".join(chunk_summaries)
    
    print("Chunk summaries combined:")
    print(f"Combined length: {len(combined.split())} words")
    print(f"\n--- COMBINED CHUNK SUMMARIES ---\n")
    print(textwrap.fill(combined, width=80))
    
    print("\n\nGenerating final summary...")
    
    summary = summarizer(
        combined,
        max_length=150,
        min_length=50,
        do_sample=False
    )
    
    return summary[0]['summary_text']

result = final_summary(chunk_summaries, summarizer)

print("\n\n=== FINAL SUMMARY ===\n")
print(textwrap.fill(result, width=80))

Chunk summaries combined:
Combined length: 297 words

--- COMBINED CHUNK SUMMARIES ---

There are no generally well-known listings for Engineers. It is possible to fall
back on the possibly indelicate and often scatological urban dictionary. Three
physicists and three Engineers were on a train to a big conference. They watched
in confusion as only one Engineer went to the ticket window and purchased a
ticket. A few minutes later it left and the conductor started collecting
tickets. The physicists were amazed to see all three Engineers get up and The
following is an attempt to categorize and collectively identify these amazing
groups of smart and talented people. There appear to be no globally recognized
official terms for groups of these people. What Do You Call A Group Of Civil
Engineers? A Foundation of Engineers. A Group of Civil Engineers is a group of
civil engineers. What Do You Call A Group Of Electrical Engineers? The sparkies
are lining up for this one. Even the Simpsons contr

In [11]:
def chunk_by_paragraph(text, max_chunk_words=150):
    """
    Split text by paragraphs rather than character count.
    Respects the natural structure of the content.
    
    Args:
        text: Full post text
        max_chunk_words: Max words before starting a new chunk
    
    Returns:
        List of chunks, each containing complete paragraphs
    """
    # Split on double newlines (paragraph breaks)
    paragraphs = [p.strip() for p in text.split('\n\n') if p.strip()]
    
    chunks = []
    current_chunk = []
    current_word_count = 0
    
    for paragraph in paragraphs:
        paragraph_words = len(paragraph.split())
        
        # If adding this paragraph exceeds limit, save current chunk and start new one
        if current_word_count + paragraph_words > max_chunk_words and current_chunk:
            chunks.append('\n\n'.join(current_chunk))
            current_chunk = [paragraph]
            current_word_count = paragraph_words
        else:
            current_chunk.append(paragraph)
            current_word_count += paragraph_words
    
    # Don't forget the last chunk
    if current_chunk:
        chunks.append('\n\n'.join(current_chunk))
    
    return chunks

# Rechunk by paragraph
para_chunks = chunk_by_paragraph(sample_post, max_chunk_words=150)

print(f"Post split into {len(para_chunks)} paragraph-based chunks")
for i, chunk in enumerate(para_chunks):
    print(f"Chunk {i+1}: {len(chunk.split())} words")

Post split into 10 paragraph-based chunks
Chunk 1: 137 words
Chunk 2: 144 words
Chunk 3: 141 words
Chunk 4: 136 words
Chunk 5: 108 words
Chunk 6: 133 words
Chunk 7: 144 words
Chunk 8: 124 words
Chunk 9: 148 words
Chunk 10: 116 words


In [12]:
# Summarize the paragraph-based chunks
para_chunk_summaries = summarize_chunks(para_chunks, summarizer)

print(f"\n✓ Generated {len(para_chunk_summaries)} chunk summaries")

Summarizing chunk 1 of 10...
  Done: There is shockingly little in the way of generally agreed upon group nomenclatur...
Summarizing chunk 2 of 10...
  Done: Three physicists and three Engineers found themselves on a train platform headin...
Summarizing chunk 3 of 10...
  Done: The following is an attempt to categorize and collectively identify these amazin...
Summarizing chunk 4 of 10...
  Done: Chemical engineers are collectively known as a Compound of Engineers. Blueprints...
Summarizing chunk 5 of 10...
  Done: A group of Civil Engineers is:A Foundation of Engineers. It would be remiss of u...
Summarizing chunk 6 of 10...
  Done: The Plug of Engineers is the name of a group of Electrical Engineers. The term n...
Summarizing chunk 7 of 10...
  Done: A good group name provides easy context for the uninitiated and for those in the...
Summarizing chunk 8 of 10...
  Done: Software Engineering contains a great many concepts related to gathering togethe...
Summarizing chunk 9 of 10...
  D

In [13]:
# Generate the final summary from paragraph-based chunks
para_result = final_summary(para_chunk_summaries, summarizer)

print("\n\n=== FINAL SUMMARY (paragraph-based chunking) ===\n")
print(textwrap.fill(para_result, width=80))

Chunk summaries combined:
Combined length: 310 words

--- COMBINED CHUNK SUMMARIES ---

There is shockingly little in the way of generally agreed upon group
nomenclature for some of the finest groups of humans to ever walk this earth.
Three physicists and three Engineers found themselves on a train platform
heading to a big conference. The physicists watched in confusion as only one
Engineer went to the ticket window and purchased a ticket. The following is an
attempt to categorize and collectively identify these amazing groups of smart
and talented people. Dealing with molecules, materials, and moles, chemical
engineers convert raw materials into usable stuff. Chemical engineers are
collectively known as a Compound of Engineers. Blueprints cover a bunch of other
sub-disciplines, from structural engineers to municipal engineering. A group of
Civil Engineers is:A Foundation of Engineers. It would be remiss of us to not
take a moment to reflect on the shear joy of this conservative force

## Findings & Analysis

### What Worked
- Paragraph-based chunking preserved complete thoughts and produced coherent 
  individual chunk summaries
- The combined chunk summaries accurately represent the full post content
- The model correctly identified the key thesis: proposing collective nouns 
  for engineering disciplines

### What Didn't Work
- The second-pass compression (summarizing the summaries) loses too much 
  context and nuance
- BART was fine-tuned on CNN/DailyMail news articles which follow a strict 
  inverted-pyramid structure — creative, humorous, narrative writing is 
  outside its training distribution
- The humor and voice of the original post is completely lost

### Key Insight
For this type of content, the **combined chunk summaries may be more useful 
than a single final summary**. Rather than forcing one final compression pass,
presenting section-by-section summaries preserves more meaning.

## Step 7: Final Summarizer — Section-by-Section Output

Based on our analysis, forcing a second compression pass loses too much context.
The most useful output for a narrative blog post is a structured 
section-by-section summary that preserves the flow of the original content.

We'll wrap everything into a single clean function that represents our finished 
summarizer.

In [14]:
def summarize_blog_post(text, summarizer, max_chunk_words=150):
    """
    Summarize a blog post using paragraph-based chunking.
    
    Rather than forcing a second compression pass (which loses context),
    this function returns structured section summaries that preserve the
    flow and meaning of the original post.
    
    Args:
        text: Full blog post text
        summarizer: Hugging Face summarization pipeline
        max_chunk_words: Max words per chunk (default 150)
    
    Returns:
        dict containing:
            - chunk_summaries: List of section summaries
            - stats: Basic statistics about the post
    """
    # Basic stats
    stats = {
        'word_count': len(text.split()),
        'char_count': len(text),
        'chunk_count': 0
    }
    
    # Chunk by paragraph
    chunks = chunk_by_paragraph(text, max_chunk_words)
    stats['chunk_count'] = len(chunks)
    
    # Summarize each chunk
    print(f"Summarizing {len(chunks)} sections...")
    summaries = []
    
    for i, chunk in enumerate(chunks):
        print(f"  Section {i+1} of {len(chunks)}...")
        
        if len(chunk.split()) < 30:
            continue
            
        summary = summarizer(
            chunk,
            max_length=60,
            min_length=20,
            do_sample=False
        )
        
        summaries.append(summary[0]['summary_text'])
    
    return {
        'chunk_summaries': summaries,
        'stats': stats
    }


# Run it
result = summarize_blog_post(sample_post, summarizer)

# Display results cleanly
print("\n")
print("=" * 60)
print("BLOG POST SUMMARY REPORT")
print("=" * 60)
print(f"Word count:    {result['stats']['word_count']}")
print(f"Sections:      {result['stats']['chunk_count']}")
print("=" * 60)

for i, summary in enumerate(result['chunk_summaries']):
    print(f"\nSection {i+1}:")
    print(textwrap.fill(summary, width=60, initial_indent="  "))

print("\n" + "=" * 60)

Summarizing 10 sections...
  Section 1 of 10...
  Section 2 of 10...
  Section 3 of 10...
  Section 4 of 10...
  Section 5 of 10...
  Section 6 of 10...
  Section 7 of 10...
  Section 8 of 10...
  Section 9 of 10...
  Section 10 of 10...


BLOG POST SUMMARY REPORT
Word count:    1331
Sections:      10

Section 1:
  There is shockingly little in the way of generally agreed
upon group nomenclature for some of the finest groups of
humans to ever walk this earth.

Section 2:
  Three physicists and three Engineers found themselves on a
train platform heading to a big conference. The physicists
watched in confusion as only one Engineer went to the ticket
window and purchased a ticket.

Section 3:
  The following is an attempt to categorize and collectively
identify these amazing groups of smart and talented people.
Dealing with molecules, materials, and moles, chemical
engineers convert raw materials into usable stuff.

Section 4:
  Chemical engineers are collectively known as a Compound of


## Conclusions & Next Steps

### What We Built
A blog post summarizer using Facebook's BART model with paragraph-based 
chunking that produces structured section-by-section summaries.

### Key Learnings
- Naive truncation produces misleading summaries for long-form content
- Character-based chunking breaks mid-sentence and confuses the model
- Paragraph-based chunking respects natural content boundaries and produces
  significantly better results
- For creative/narrative writing, section-by-section summaries outperform
  forced single-paragraph compression
- BART performs best on factual, news-style content — creative and humorous 
  writing is outside its training distribution

### Limitations
- BART was fine-tuned on CNN/DailyMail news articles — it loses the voice 
  and humor of creative writing
- Section summaries are useful but not a true abstractive summary of the 
  whole post
- Model inference is slow on CPU (~5-7 seconds per chunk)

### Potential Next Steps
- Test alternative models (PEGASUS, T5) on the same content for comparison
- Fine-tune BART on a custom dataset of blog posts for better style matching
- Build a simple Gradio UI to make the summarizer usable without the notebook
- Automate TL;DR generation and inject it at the top of each published post
- Evaluate summary quality systematically across multiple posts

## Appendix: Loading Posts From Files

To summarize a new blog post, save it as a `.txt` file in the `notebooks/` 
folder and run the cell below.

In [ ]:
def load_post_from_file(filename):
    """
    Load a blog post from a text file in the notebooks directory.
    
    Args:
        filename: Name of the .txt file (e.g. 'my-post.txt')
    
    Returns:
        Post text as a string
    """
    filepath = os.path.join(os.getcwd(), filename)
    
    if not os.path.exists(filepath):
        raise FileNotFoundError(f"Could not find {filepath}")
    
    with open(filepath, 'r', encoding='utf-8') as f:
        text = f.read()
    
    print(f"Loaded: {filename}")
    print(f"Word count: {len(text.split())}")
    
    return text

# Example usage — uncomment and swap in your filename:
# post = load_post_from_file('my-post.txt')
# result = summarize_blog_post(post, summarizer)